File: leetcode/1206-design-skiplist/solution.py
import random


class _Node:
    __slots__ = ("val", "forward")

    def __init__(self, val, level):
        self.val = val
        self.forward = [None] * (level + 1)


class Skiplist:
    MAX_LEVEL = 16
    P = 0.5

    def __init__(self):
        self.head = _Node(-1, self.MAX_LEVEL)
        self.level = 0

    def _random_level(self):
        lvl = 0
        while random.random() < self.P and lvl < self.MAX_LEVEL:
            lvl += 1
        return lvl

    def _find_predecessors(self, target):
        # for each level, the last node whose value is strictly less than target
        update = [self.head] * (self.MAX_LEVEL + 1)
        cur = self.head
        for i in range(self.level, -1, -1):
            while cur.forward[i] and cur.forward[i].val < target:
                cur = cur.forward[i]
            update[i] = cur
        return update

    def search(self, target: int) -> bool:
        update = self._find_predecessors(target)
        candidate = update[0].forward[0]
        return candidate is not None and candidate.val == target

    def add(self, num: int) -> None:
        update = self._find_predecessors(num)
        new_level = self._random_level()
        if new_level > self.level:
            for i in range(self.level + 1, new_level + 1):
                update[i] = self.head
            self.level = new_level
        new_node = _Node(num, new_level)
        for i in range(new_level + 1):
            new_node.forward[i] = update[i].forward[i]
            update[i].forward[i] = new_node

    def erase(self, num: int) -> bool:
        update = self._find_predecessors(num)
        candidate = update[0].forward[0]
        if candidate is None or candidate.val != num:
            return False
        for i in range(self.level + 1):
            if update[i].forward[i] is not candidate:
                break
            update[i].forward[i] = candidate.forward[i]
        while self.level > 0 and self.head.forward[self.level] is None:
            self.level -= 1
        return True


# Your Skiplist object will be instantiated and called as such:
# obj = Skiplist()
# param_1 = obj.search(target)
# obj.add(num)
# param_3 = obj.erase(num)

File: leetcode/1206-design-skiplist/Solution.java
import java.util.Random;

class Skiplist {
    private static final int MAX_LEVEL = 16;
    private static final double P = 0.5;

    private class Node {
        int val;
        Node[] forward;
        Node(int val, int level) {
            this.val = val;
            this.forward = new Node[level + 1];
        }
    }

    private final Node head = new Node(-1, MAX_LEVEL);
    private int level = 0;
    private final Random random = new Random();

    private int randomLevel() {
        int lvl = 0;
        while (random.nextDouble() < P && lvl < MAX_LEVEL) {
            lvl++;
        }
        return lvl;
    }

    @SuppressWarnings("unchecked")
    private Node[] findPredecessors(int target) {
        Node[] update = new Node[MAX_LEVEL + 1];
        Node cur = head;
        for (int i = level; i >= 0; i--) {
            while (cur.forward[i] != null && cur.forward[i].val < target) {
                cur = cur.forward[i];
            }
            update[i] = cur;
        }
        return update;
    }

    public boolean search(int target) {
        Node[] update = findPredecessors(target);
        Node candidate = update[0].forward[0];
        return candidate != null && candidate.val == target;
    }

    public void add(int num) {
        Node[] update = findPredecessors(num);
        int newLevel = randomLevel();
        if (newLevel > level) {
            for (int i = level + 1; i <= newLevel; i++) {
                update[i] = head;
            }
            level = newLevel;
        }
        Node newNode = new Node(num, newLevel);
        for (int i = 0; i <= newLevel; i++) {
            newNode.forward[i] = update[i].forward[i];
            update[i].forward[i] = newNode;
        }
    }

    public boolean erase(int num) {
        Node[] update = findPredecessors(num);
        Node candidate = update[0].forward[0];
        if (candidate == null || candidate.val != num) {
            return false;
        }
        for (int i = 0; i <= level; i++) {
            if (update[i].forward[i] != candidate) break;
            update[i].forward[i] = candidate.forward[i];
        }
        while (level > 0 && head.forward[level] == null) {
            level--;
        }
        return true;
    }
}

/**
 * Your Skiplist object will be instantiated and called as such:
 * Skiplist obj = new Skiplist();
 * boolean param_1 = obj.search(target);
 * obj.add(num);
 * boolean param_3 = obj.erase(num);
 */
